### DISCUSSION TITLE
Testing with Jest🧑‍💻

### DISCUSSION THREAD

**Discussion by LOGESH-B** [gbowne1/codebooker#167] [General]
Hii @gbowne1 
I am learning  testing with jest....
I now ready with tesing the backend apis
But when it comes for frontend I am only clear to write the test about wheather the element is present in the DOM or not
I can't able to figure out how the fronted flow is tested with jest
Can you have any tutorials or blogs or documentation for Unit test
Wheather we have any Collabarator who knows testing?

**Comment by gbowne1** [comment]
I think a good test workflow for us is to have:

1 or 2 developer testing the API
1 or 2 developer testing the UI
1 or 2 developer testing the express server
1 or 2 developer testing the database

Consider that we need to do integration tests, end-to-end (e2e) and unit tests.

We should be using the official Jest documentation for the most part to learn from, however there are a number of great tutorials.    

https://www.youtube.com/watch?v=5QRx-i5jVNw
https://www.youtube.com/watch?v=7r4xVDI2vho
https://www.youtube.com/watch?v=OVNjsIto9xM
https://www.youtube.com/watch?v=ML5egqL3YFE
https://www.youtube.com/watch?v=r9HdJ8P6GQI
https://www.youtube.com/watch?v=3e1GHCA3GP0

It might take doing a few tutorials to understand.  I would not expect a single person to do all of our testing, but it will be a big benefit.  If you are only able to test whether the element is present in the DOM, that is a good start.

If you use Visual Studio Code or Visual Studio, there is a Jest extension you should install, just search for Jest in the marketplace and in the Extension search in VSCode.

@LOGESH-B @BlackBond06 @victor20m @ratishjain12 ,etc

**Reply by gbowne1** [reply]
There are tutorials, especially on Youtube that cover the topic of integration, unit and e2e testing as an overview, not Jest specific.  I would also suggest looking at some of those.

At some point we should require all tests pass before we merge.  This will mean that before the PR gets sent, all required tests must pass, and when we review the PR, the tests must pass.  We report back to the developer doing the PR, if there are any tests not passing.  This will help us immensely with making sure that especially when we are adding new features, it will help keep from breaking changes, especially any new code that would break old code/features.

**Reply by gbowne1** [reply]
I found some more video resources:

https://www.youtube.com/watch?v=9Yrd4aZkse8
https://www.youtube.com/playlist?list=PLDoPjvoNmBAwSrfBPERTnCmWAbcMAwG9O <playlist of 20 videos on Jest>
https://www.youtube.com/watch?v=8vfQ6SWBZ-U
https://www.youtube.com/watch?v=Flo268xRpV0
https://www.youtube.com/watch?v=QzznzOlkgGw
https://www.youtube.com/playlist?list=PLC3y8-rFHvwirqe1KHFCHJ0RqNuN61SJd <Playlist of 52 videos on Jest>
https://www.youtube.com/watch?v=Jv2uxzhPFl4
https://www.youtube.com/watch?v=D9DdY2WmM-s

**Reply by LOGESH-B** [reply]
https://www.youtube.com/playlist?list=PLC3y8-rFHvwirqe1KHFCHJ0RqNuN61SJd <Playlist of 52 videos on Jest> I folllowed this playlist

**Reply by gbowne1** [reply]
it's pretty good. I am on the 4th video of the list

**Comment by BlackBond06** [comment]
Yes I agree test at this point, is very important, as the app is really growing fast. But, it may be a bit tricky seeing that some of us are still new to test. Nevertheless, I'll begin looking into testing small parts of the app to understand what it is, and ultimately make more test contributions.

**Reply by gbowne1** [reply]
I am well aware people may be new to testing. We can start out making our testing simple and grow as needs come about and as people are learning testing.  The more people we have that are able to at least understand testing the better our developer experience witll be.

I suggest going through the tutorials I included here and doing a few of their examples to get some knowledge as well as videos/tutorials about unit testing, e2e testing and integration testing.

Create React App gave us App.test.js which was meant to test the rendering of App.js, so it could ideally be used to test App.js itself.  

We should create a test folder in the most logical place, and install jest as a devDependency. with -D or --save-dev.  

I think we should also primarily be using Postman to test our API.
